p2p/host/basic/natmgr.go

```diff
@@ -87,151 +90,151 @@ func (nmgr *natManager) discoverNAT() {
-	for _, mapping := range nat.Mappings() {
-		if mapping.InternalAddr().Equal(intaddr) {
-			return // it exists! return.
+	nmgr.proc.Go(func(_ goprocess.Process) {
+		nmgr.syncMu.Lock()
+		defer nmgr.syncMu.Unlock()
+
+		ports := map[string]map[int]bool{
+			"tcp": map[int]bool{},
+			"udp": map[int]bool{},
 		}
-	}
+		for _, maddr := range nmgr.net.ListenAddresses() {
+			// Strip the IP
+			maIP, rest := ma.SplitFirst(maddr)
+			if maIP == nil || rest == nil {
+				continue
+			}
 
-	ctx := context.TODO()
-	lm := make(lgbl.DeferredMap)
-	lm["internalAddr"] = func() interface{} { return intaddr.String() }
+			switch maIP.Protocol().Code {
+			case ma.P_IP6, ma.P_IP4:
+			default:
+				continue
+			}
 
-	defer log.EventBegin(ctx, "natMgrAddPortMappingWait", lm).Done()
+			// Only bother if we're listening on a
+			// unicast/unspecified IP.
+			ip := net.IP(maIP.RawValue())
+			if !(ip.IsGlobalUnicast() || ip.IsUnspecified()) {
+				continue
+			}
 
-	select {
-	case <-nmgr.proc.Closing():
-		lm["outcome"] = "cancelled"
-		return // no use.
-	case <-nmgr.ready: // wait until it's ready.
-	}
+			// Extract the port/protocol
+			proto, _ := ma.SplitFirst(rest)
+			if proto == nil {
+				continue
+			}
 
-	// actually start the port map (sub-event because waiting may take a while)
-	defer log.EventBegin(ctx, "natMgrAddPortMapping", lm).Done()
+			var protocol string
+			switch proto.Protocol().Code {
+			case ma.P_TCP:
+				protocol = "tcp"
+			case ma.P_UDP:
+				protocol = "udp"
+			default:
+				continue
+			}
 
-	// get the nat
-	m, err := nat.NewMapping(intaddr)
-	if err != nil {
-		lm["outcome"] = "failure"
-		lm["error"] = err
-		return
-	}
+			port, err := strconv.ParseUint(proto.Value(), 10, 16)
+			if err != nil {
+				// bug in multiaddr
+				panic(err)
+			}
+			ports[protocol][int(port)] = false
+		}
 
-	extaddr, err := m.ExternalAddr()
-	if err != nil {
-		lm["outcome"] = "failure"
-		lm["error"] = err
-		return
-	}
+		var wg sync.WaitGroup
+		defer wg.Wait()
+
+		// Close old mappings
+		for _, m := range nat.Mappings() {
+			mappedPort := m.InternalPort()
+			if _, ok := ports[m.Protocol()][mappedPort]; !ok {
+				// No longer need this mapping.
+				wg.Add(1)
+				go func(m inat.Mapping) {
+					defer wg.Done()
+					m.Close()
+				}(m)
+			} else {
+				// already mapped
+				ports[m.Protocol()][mappedPort] = true
+			}
+		}
 
-	lm["outcome"] = "success"
-	lm["externalAddr"] = func() interface{} { return extaddr.String() }
-	log.Infof("established nat port mapping: %s <--> %s", intaddr, extaddr)
+		// Create new mappings.
+		for proto, pports := range ports {
+			for port, mapped := range pports {
+				if mapped {
+					continue
+				}
+				wg.Add(1)
+				go func(proto string, port int) {
+					defer wg.Done()
+					_, err := nat.NewMapping(proto, port)
+					if err != nil {
+						log.Errorf("failed to port-map %s port %d: %s", proto, port, err)
+					}
+				}(proto, port)
+			}
+		}
+	})
```

Comment:
Not relevant to the changes in this PR, but shouldn't we be triggering an "identify push" here if there were changes (maybe not _exactly_ here, but when the mappings are done)?

Reply:
Probably (https://github.com/libp2p/go-libp2p/issues/553). However, we also need to push when:

1. Our observed addresses change significantly.
2. Our interface addresses change.

etc...

Reply:
Eventbus 🚌🚌🚌